packages/flutter/lib/src/widgets/display_feature_sub_screen.dart

```diff
@@ -0,0 +1,233 @@
+    final TextDirection textDirection = Directionality.of(context);
+    switch (textDirection) {
+      case TextDirection.rtl:
+        return const Offset(double.maxFinite, 0);
+      case TextDirection.ltr:
+        return Offset.zero;
+    }
+  }
+
+  static Iterable<Rect> _avoidBounds(MediaQueryData mediaQuery) {
+    return mediaQuery.displayFeatures.map((DisplayFeature d) => d.bounds)
+        .where((Rect r) => r.shortestSide > 0);
+  }
+
+  /// Returns the closest sub-screen to the [anchorPoint]
+  static Rect _closestToAnchorPoint(Iterable<Rect> subScreens, Offset anchorPoint) {
+    return subScreens.fold(subScreens.first,
+        (Rect previousValue, Rect element) {
+      final double previousDistance =
+          _distanceFromPointToRect(anchorPoint, previousValue);
+      final double elementDistance =
+          _distanceFromPointToRect(anchorPoint, element);
+      if (previousDistance < elementDistance)
+        return previousValue;
+      else
+        return element;
+    });
+  }
+
+  static double _distanceFromPointToRect(Offset point, Rect rect) {
+    // Cases for point position relative to rect:
+    // 1  2  3
+    // 4 [R] 5
+    // 6  7  8
+    if (point.dx < rect.left) {
+      if (point.dy < rect.top) {
+        // Case 1
+        return (point - rect.topLeft).distance;
+      } else if (point.dy > rect.bottom) {
+        // Case 6
+        return (point - rect.bottomLeft).distance;
+      } else {
+        // Case 4
+        return rect.left - point.dx;
+      }
+    } else if (point.dx > rect.right) {
+      if (point.dy < rect.top) {
+        // Case 3
+        return (point - rect.topRight).distance;
+      } else if (point.dy > rect.bottom) {
+        // Case 8
+        return (point - rect.bottomRight).distance;
+      } else {
+        // Case 5
+        return point.dx - rect.right;
+      }
+    } else {
+      if (point.dy < rect.top) {
+        // Case 2
+        return rect.top - point.dy;
+      } else if (point.dy > rect.bottom) {
+        // Case 7
+        return point.dy - rect.bottom;
+      } else {
+        // Case R
+        return 0;
+      }
+    }
+  }
+
+  /// Returns sub-screens resulted by dividing [wantedBounds] along items of
+  /// [avoidBounds] that are at least as high or as wide.
+  static Iterable<Rect> _subScreensInBounds(Rect wantedBounds, Iterable<Rect> avoidBounds) {
+    Iterable<Rect> subScreens = <Rect>[wantedBounds];
+    for (final Rect bounds in avoidBounds) {
+      subScreens = subScreens.expand((Rect screen) {
+        final List<Rect> results = <Rect>[];
+        if (screen.top >= bounds.top && screen.bottom <= bounds.bottom) {
+          // Display feature splits the screen vertically
+          if (screen.left < bounds.left) {
+            // There is a smaller sub-screen, left of the display feature
+            results.add(Rect.fromLTWH(screen.left, screen.top,
+                bounds.left - screen.left, screen.height));
```

Comment:
formatting nit (here and below):

```suggestion
            results.add(Rect.fromLTWH(
              screen.left,
              screen.top,
              bounds.left - screen.left,
              screen.height,
            ));
```